src/Compilers/CSharp/Portable/CSharpResources.resx

```diff
@@ -6600,4 +6600,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_FunctionPointerTypesInAttributeNotSupported" xml:space="preserve">
     <value>Using a function pointer type in a 'typeof' in an attribute is not supported.</value>
   </data>
+  <data name="IDS_FeatureSealedToStringInRecord" xml:space="preserve">
+    <value>sealed ToString in record</value>
+  </data>
+  <data name="ERR_InheritingFromRecordWithSealedToString" xml:space="preserve">
+    <value>Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater.</value>
```

Comment:
>Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater. [](start = 11, length = 132)

We don't hardcode language versions for errors like this. For example, see how we encorpoarte language version information into ERR_InvalidModifierForLanguageVersion.  
We also need to make sure that IDE will offer upgrading the language version for the progect based on the error, it looks like some changes should be made in src\Features\CSharp\Portable\UpgradeProject\CSharpUpgradeProjectCodeFixProvider.cs to make this work. Appropriate test(s) should be added in src\EditorFeatures\CSharpTest\Diagnostics\UpgradeProject\UpgradeProjectTests.cs to verify the fact.

Reply:
I changed the error message to use placeholders, but I haven't made the changes on project upgrade yet.

Reply:
> We also need to make sure that IDE will offer upgrading the language version for the progect based on the error, it looks like some changes should be made in src\Features\CSharp\Portable\UpgradeProject\CSharpUpgradeProjectCodeFixProvider.cs to make this work. Appropriate test(s) should be added in src\EditorFeatures\CSharpTest\Diagnostics\UpgradeProject\UpgradeProjectTests.cs to verify the fact.

This is proving more difficult than I anticipated... The implementation is trivial (just add CS8912 to the `FixableDiagnosticIds` property), but I'm unable to test it.

There are 2 errors related to this feature:
- CS8652 (ERR_FeatureInPreview), which is already covered by the tests, but isn't specific to this feature
- CS8912 (ERR_InheritingFromRecordWithSealedToString)

In order to trigger CS8912, the base record must be declared in a different module, but the current test infrastructure doesn't allow for that. I'd change it, but I'm not familiar with the workspace API and I don't really know where to start...

Reply:
> In order to trigger CS8912, the base record must be declared in a different module, but the current test infrastructure doesn't allow for that. I'd change it, but I'm not familiar with the workspace API and I don't really know where to start...

Actually it looks like test infatrsucture supports testing scenarios when more than one compilation is involved. See UpgradeProjectForDefaultInterfaceImplementation_CS8706 for example.

---
In reply to: [605998135](https://github.com/dotnet/roslyn/pull/52029#discussion_r605998135) [](ancestors = 605998135)

Reply:
Thanks, I'm looking into it

Reply:
OK, done